external-crates/move/crates/move-compiler/src/expansion/name_validation.rs

```diff
@@ -0,0 +1,378 @@
+
+// Implicit aliases for the Move Stdlib:
+// use std::vector;
+// use std::option::{Self, Option};
+pub const IMPLICIT_STD_MODULES: &[Symbol] = &[symbol!("option"), symbol!("vector")];
+pub const IMPLICIT_STD_MEMBERS: &[(Symbol, Symbol, ModuleMemberKind)] = &[(
+    symbol!("option"),
+    symbol!("Option"),
+    ModuleMemberKind::Struct,
+)];
+
+// Implicit aliases for Sui mode:
+// use sui::object::{Self, ID, UID};
+// use sui::transfer;
+// use sui::tx_context::{Self, TxContext};
+pub const IMPLICIT_SUI_MODULES: &[Symbol] = &[
+    symbol!("object"),
+    symbol!("transfer"),
+    symbol!("tx_context"),
+];
+pub const IMPLICIT_SUI_MEMBERS: &[(Symbol, Symbol, ModuleMemberKind)] = &[
+    (symbol!("object"), symbol!("ID"), ModuleMemberKind::Struct),
+    (symbol!("object"), symbol!("UID"), ModuleMemberKind::Struct),
+    (
+        symbol!("tx_context"),
+        symbol!("TxContext"),
+        ModuleMemberKind::Struct,
+    ),
+];
+
+#[derive(Copy, Clone, Debug)]
+pub enum ModuleMemberKind {
+    Constant,
+    Function,
+    Struct,
+    Enum,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum NameCase {
+    Constant,
+    Function,
+    Struct,
+    Enum,
+    Module,
+    ModuleMemberAlias(ModuleMemberKind),
+    ModuleAlias,
+    Variable,
+    Address,
+    TypeParameter,
+}
+
+//**************************************************************************************************
+// impls
+//**************************************************************************************************
+
+impl ModuleMemberKind {
+    pub fn case(self) -> NameCase {
+        match self {
+            ModuleMemberKind::Constant => NameCase::Constant,
+            ModuleMemberKind::Function => NameCase::Function,
+            ModuleMemberKind::Struct => NameCase::Struct,
+            ModuleMemberKind::Enum => NameCase::Enum,
+        }
+    }
+}
+
+impl NameCase {
+    pub const fn name(&self) -> &'static str {
+        match self {
+            NameCase::Constant => "constant",
+            NameCase::Function => "function",
+            NameCase::Struct => "struct",
+            NameCase::Enum => "enum",
+            NameCase::Module => "module",
+            NameCase::ModuleMemberAlias(ModuleMemberKind::Function) => "function alias",
+            NameCase::ModuleMemberAlias(ModuleMemberKind::Constant) => "constant alias",
+            NameCase::ModuleMemberAlias(ModuleMemberKind::Struct) => "struct alias",
+            NameCase::ModuleMemberAlias(ModuleMemberKind::Enum) => "enum alias",
+            NameCase::ModuleAlias => "module alias",
+            NameCase::Variable => "variable",
+            NameCase::Address => "address",
+            NameCase::TypeParameter => "type parameter",
+        }
+    }
+}
+
+//**************************************************************************************************
+// Valid names
+//**************************************************************************************************
+
```

Comment:
It's strange to me to change these all to `Result<(),Box<Diagnostic>>` from their original interface. Instead they should take `CompilationEnv` and retain their previous behavior.

Reply:
I was inspired by how parser does things but admittedly passing environment is simpler and leads to fewer code changes so ... done!